control_plane/attachment_service/src/service.rs

```diff
@@ -3970,6 +3970,28 @@ impl Service {
         reconciles_spawned
     }
 
+    /// Useful for tests: run whatever work a background [`Self::reconcile_all`] would have done, but
+    /// also wait for any generated Reconcilers to complete.  Calling this until it returns zero should
+    /// put the system into a quiescent state where future background reconciliations won't do anything.
+    pub(crate) async fn reconcile_all_now(&self) -> Result<usize, ReconcileWaitError> {
+        self.reconcile_all();
+
+        let waiters = {
+            let mut waiters = Vec::new();
+            let locked = self.inner.read().unwrap();
+            for (_tenant_shard_id, shard) in locked.tenants.iter() {
+                if let Some(waiter) = shard.get_waiter() {
+                    waiters.push(waiter);
+                }
+            }
+            waiters
+        };
+
+        let waiter_count = waiters.len();
+        self.await_waiters(waiters, RECONCILE_TIMEOUT).await?;
+        Ok(waiter_count)
```

Comment:
We return the nonzero value because if the waiters complete, it doesn't mean we're quiescent: tenants can require multiple rounds of reconciliation.  The main case is during a split:
- One round of optimization to move the attached location away from the split location
- Second round of optimization to move the now-secondary location away from the split location.

It's simpler to do this as separate optimization steps than to try  and teach reconciler how to gracefully do both in one step.

Reply:
What you're saying makes sense, but I still don't see the point awaiting the waiters here.

Reply:
When we use this in a loop from a test, the alternative would be either to busy-spin or have some arbitrary wait between checks -- the wait enables the tests to wait exactly as long as it takes to do the reconcile